sql/hive/src/test/scala/org/apache/spark/sql/hive/execution/HiveSemiJoinSuite.scala

```diff
@@ -0,0 +1,450 @@
+      |LIMIT 5""".stripMargin)
+
+  createQueryTest(
+    "(correlated)WHERE clause with NOT IN #1",
+    """select *
+    |from src b
+    |where b.key not in
+    |        (select a.key
+    |         from src a
+    |         where b.value = a.value and a.key > 9
+    |        )
+    |order by key, value
+    |LIMIT 5"""
+      .stripMargin)
+
+  createQueryTest(
+    "(correlated)WHERE clause with IN #2",
+    """select *
+    |from src b
+    |where b.key in
+    |        (select a.key
+    |         from src a
+    |         where b.value = a.value and a.key > 9
+    |        )
+    |and b.key > 15
+    |order by key, value
+    |LIMIT 5""".stripMargin)
+
+  createQueryTest(
+    "(correlated)WHERE clause with NOT IN #2",
+    """select *
+    |from src b
+    |where b.key not in
+    |        (select a.key
+    |         from src a
+    |         where b.value = a.value and a.key > 9
+    |        )
+    |and b.key > 15
+    |order by key, value
+    |LIMIT 5""".
+      stripMargin)
+
+  createQueryTest(
+    "(correlated)WHERE clause with EXISTS #1",
+    """select *
+    |from src b
+    |where EXISTS
+    |        (select a.key
+    |         from src a
+    |         where b.value = a.value and a.key > 9
+    |        )
+    |order by key, value
+    |LIMIT 5""".
+      stripMargin)
+
+  createQueryTest(
+    "(correlated)WHERE clause with NOT EXISTS #1",
+    """select *
+    |from src b
+    |where NOT EXISTS
+    |        (select a.key
+    |         from src a
+    |         where b.value = a.value and a.key > 9
+    |        )
+    |order by key, value
+    |LIMIT 5""".stripMargin)
+
+  createQueryTest(
+    "(correlated)WHERE clause with EXISTS #2",
+    """select *
+    |from src b
+    |where EXISTS
+    |        (select a.key
+    |         from src a
+    |         where b.value = a.value and a.key > 9
+    |        )
+    |and b.key > 15
+    |order by key, value
+    |LIMIT 5""".
+      stripMargin)
+
+  createQueryTest(
+    "(correlated)WHERE clause with NOT EXISTS #2",
+    """select *
+    |from src b
+    |where NOT EXISTS
+    |        (select a.key % 291
+    |         from src a
+    |         where b.value = a.value and a.key > 9
+    |        )
+    |and b.key > 15
+    |order by key, value
+    |LIMIT 5""".
+      stripMargin)
+}
+
+class SemiJoinHashJoin extends HiveSemiJoinSuite {
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    enableSortMerge(false)
+    enableTungsten(false)
+    enableCodeGen(false)
+  }
+}
+
+class SemiJoinHashJoinTungsten extends HiveSemiJoinSuite {
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    enableSortMerge(false)
+    enableTungsten(true)
+    enableCodeGen(true)
+  }
+}
+
+class SemiJoinSortMerge extends HiveSemiJoinSuite {
+  override def beforeAll(): Unit = {
+    super.beforeAll()
+    enableSortMerge(true)
+    enableTungsten(false)
+    enableCodeGen(false)
+  }
+}
+
+class SemiJoinSortMergeTungsten extends HiveSemiJoinSuite {
```

Comment:
Oh, even I am not so sure, but more unit test is harmless, just in cases people will add the sort merge semi join support in the future.